File: javascript/easy/implementQueueUsingStacks.js
/**
 * Initialize your data structure here.
 */
var MyQueue = function() {
    this.pushStack = [];
    this.popStack = [];
};

/**
 * Push element x to the back of queue. 
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function(x) {
    this.pushStack.push(x);
};

/**
 * Removes the element from in front of queue and returns that element.
 * @return {number}
 */
MyQueue.prototype.pop = function() {
    if(this.popStack.length <= 0) {
        while(this.pushStack.length > 0) {
            this.popStack.push(this.pushStack.pop());
        }
    }
    
    return this.popStack.pop();
};

/**
 * Get the front element.
 * @return {number}
 */
MyQueue.prototype.peek = function() {
    if(this.popStack.length <= 0) {
        while(this.pushStack.length > 0) {
            this.popStack.push(this.pushStack.pop());
        }
    }
    return this.popStack[this.popStack.length-1];
};

/**
 * Returns whether the queue is empty.
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {
    return (this.popStack.length == 0) && (this.pushStack.length == 0);
};

/** 
 * Your MyQueue object will be instantiated and called as such:
 * var obj = Object.create(MyQueue).createNew()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */

File: ruby/easy/implement_queue_using_stacks.rb
class MyQueue

=begin
    Initialize your data structure here.
=end
    def initialize()
        @push_stack = []
        @pop_stack = []
    end


=begin
    Push element x to the back of queue.
    :type x: Integer
    :rtype: Void
=end
    def push(x)
        @push_stack.push(x)
    end


=begin
    Removes the element from in front of queue and returns that element.
    :rtype: Integer
=end
    def pop()
        copy(@push_stack, @pop_stack) if @pop_stack.empty?
        @pop_stack.pop
    end


=begin
    Get the front element.
    :rtype: Integer
=end
    def peek()
        copy(@push_stack, @pop_stack) if @pop_stack.empty?
        @pop_stack.last
    end


=begin
    Returns whether the queue is empty.
    :rtype: Boolean
=end
    def empty()
        @push_stack.empty? && @pop_stack.empty?
    end

    private
    def copy(from, to)
        while(!from.empty?)
            to.push(from.pop)
        end
    end
    
end

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue.new()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()